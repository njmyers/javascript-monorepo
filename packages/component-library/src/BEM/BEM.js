// @flow
import * as React from 'react';

type Props = {
  children: React.Node,
  block: string,
};

type ChildProps = {
  className?: string,
  element?: string | void,
  modifiers?: string | void | Array<string>,
};

class BEM extends React.Component<Props> {
  static defaultProps = {
    children: null,
    block: '',
  };

  // coerce modifiers to array with at least one empty string
  typeCheckModifiers = (modifiers: string | Array<string> = '') => {
    return Array.isArray(modifiers)
      ? ['', ...modifiers]
      : modifiers
        ? ['', modifiers]
        : [''];
  };

  // coerce element to empty string
  typeCheckElement = (element: string = '') => element;

  /** Calculates correct classNames
    1. block_element-modifier
    2. block-modifier
    3. element
    4. element-modifier
  */
  getClassName = (block: string, element: string, modifier: string) => {
    return block
      ? element
        ? modifier
          ? `${block}_${element}-${modifier}`
          : `${block}_${element}`
        : modifier
          ? `${block}-${modifier}`
          : block
      : modifier
        ? `${element}-${modifier}`
        : element;
  };

  getAllClasses = (element: string, modifiers: Array<string>) => {
    return modifiers
      .map((modifier) => this.getClassName(this.props.block, element, modifier))
      .reduce((a, b) => (a ? `${a} ${b}` : b), '');
  };

  /**
   * Merges existing classNames with new generated classNames
   * @param  {[type]} props [description]
   * @return {[type]}       [description]
   */
  mergeClassNames = (props: ChildProps) => {
    const safeElement = this.typeCheckElement(props.element);
    const safeModifiers = this.typeCheckModifiers(props.modifiers);

    const newClasses = this.getAllClasses(safeElement, safeModifiers);

    const className = props.className
      ? `${props.className} ${newClasses}`
      : newClasses;

    return { className };
  };

  /**
   * Checks whether react element is of DOM type
   * @param  {[type]}  type [description]
   * @return {Boolean}      [description]
   */
  isDOMElement = ({ type }: React.Element<any>): boolean => {
    return Boolean(typeof type === 'string');
  };

  /**
   * Checks if react element has children
   * @param  {[type]}  props [description]
   * @return {Boolean}       [description]
   */
  hasChildren = ({ props }: React.Element<any>): boolean => {
    return Boolean(props && props.children);
  };

  /**
   * Deeply maps react children but only if they are valid DOM elements
   * @param  {[type]} children  [description]
   * @param  {[type]} deepMapFn [description]
   * @return {[type]}           [description]
   */
  deepDOMMap = (children: React.Node, deepMapFn: (child: Node) => Node) => {
    return React.Children.map(children, (child) => {
      return React.isValidElement(child)
        ? // if not dom element do not traverse
          this.isDOMElement(child)
          ? this.hasChildren(child)
            ? deepMapFn(
                React.cloneElement(
                  child,
                  child.props,
                  this.deepDOMMap(child.props.children, deepMapFn)
                )
              )
            : deepMapFn(child)
          : child
        : child;
    });
  };

  render() {
    return this.deepDOMMap(this.props.children, (child) => {
      return React.cloneElement(child, {
        ...child.props,
        ...this.mergeClassNames(child.props),
      });
      return child;
    });
  }
}

export default BEM;
