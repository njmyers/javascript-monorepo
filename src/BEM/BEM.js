// @flow
import * as React from 'react';
import Children from 'react-children-utilities';

type Props = {
  children: React.Node,
  block: string,
};

type ChildProps = {
  className?: string,
  element?: string | void,
  modifiers?: string | void | Array<string>,
};

class BEM extends React.PureComponent<Props> {
  static defaultProps = {
    children: null,
    block: '',
  };

  // coerce modifiers to array with at least one empty string
  typeCheckModifiers = (modifiers: string | Array<string> = '') => {
    return Array.isArray(modifiers)
      ? ['', ...modifiers]
      : modifiers
        ? ['', modifiers]
        : [''];
  };

  // coerce element to empty string
  typeCheckElement = (element: string = '') => element;

  /** Calculates correct classNames
    1. block_element-modifier
    2. block-modifier
    3. element
    4. element-modifier
  */
  getClassName = (block: string, element: string, modifier: string) => {
    return block
      ? element
        ? modifier
          ? `${block}_${element}-${modifier}`
          : `${block}_${element}`
        : modifier
          ? `${block}-${modifier}`
          : block
      : modifier
        ? `${element}-${modifier}`
        : element;
  };

  getAllClasses = (element: string, modifiers: Array<string>) => {
    return modifiers
      .map((modifier) => this.getClassName(this.props.block, element, modifier))
      .reduce((a, b) => (a ? `${a} ${b}` : b), '');
  };

  mergeClassNames = (props: ChildProps) => {
    const safeElement = this.typeCheckElement(props.element);
    const safeModifiers = this.typeCheckModifiers(props.modifiers);

    const newClasses = this.getAllClasses(safeElement, safeModifiers);

    const className = props.className
      ? `${props.className} ${newClasses}`
      : newClasses;

    return { className };
  };

  render() {
    return Children.deepMap(this.props.children, (element: React.Node) => {
      return React.isValidElement(element)
        ? typeof element.type !== 'function'
          ? React.cloneElement(element, this.mergeClassNames(element.props))
          : React.cloneElement(element)
        : null;
    });
  }
}

export default BEM;
