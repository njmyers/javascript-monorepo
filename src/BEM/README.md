## BEM

### Background (Frustration)

I've read many articles about using BEM syntax. As I understand it the main points are as follows.

1.  BEM syntax reduces css specificity (GOOD)
2.  BEM syntax can be implemented easily with preprocessors (GOOD)
3.  BEM syntax makes your HTML or Components look ugly (BAD!)

If you stick to BEM syntax then you end up with a React Component like this

```js
<p className="block-element block_element-modifier block_element-otherModifier">
  Text
</p>
```

And sass or scss that looks like this

```scss
.block
  &_element
    &-modifier
    &-otherModifier
```

The sass looks like great IMHO but the markup looks terrible. How you want to write your sass or scss is a topic for another conversation. All of the aforementioned articles then proceed to explain how they have made adjustments on the BEM syntax so that it is easier to read. Usually this means creating shorter classNames and using tricks.

TODO ADD EXAMPLE

```js
```

```scss
```

Now of course this is a good thing as we want our source code to be cleaner and easier to read. However it comes at the cost of possible reducing the specificity of your compilied css code. Why do we have to make a sacrifice?

### Solution

Now we have a brand new component for writing BEM classNames to your components. We hopefully get the best of both worlds by being able to using the full power of BEM in terms of specificity while using a much cleaner syntax in React. Behold the power of `<BEM>`

```js
class Menu extends React.Component {
  static defaultProps = {
    menuItems: [
      {
        text: 'Link1',
        id: 1,
      },
      {
        text: 'Link2',
        id: 2,
      },
      {
        text: 'Link3',
        id: 3,
      },
    ],
  };

  state = {
    active: 1,
  };

  render() {
    return (
      <BEM block="menu">
        <nav>
          {this.props.menuItems.map((item) => (
            <div element="buttonContainer">
              <button
                element="button"
                modifiers={this.state.active === item.id ? 'active' : ''}
              >
                {item.text}
              </button>
            </div>
          ))}
        </nav>
      </BEM>
    );
  }
}
```

Now what will this render? Something like this of course depending on which data you pass in

```html
<nav class="menu">
  <div element="buttonContainer" class="menu_buttonContainer">
    <button element="button" modifiers="active" class="menu_button menu_button-active">Link1</button>
  </div
  ><div element="buttonContainer" class="menu_buttonContainer">
    <button element="button" modifiers="" class="menu_button">Link2</button>
  </div>
  <div element="buttonContainer" class="menu_buttonContainer">
    <button element="button" modifiers="" class="menu_button">Link3</button>
  </div>
</nav>
```

Great... not only do we have the wordy classes all properly applied we also have searchable property strings associated with each element. Yay success

How would it look manually applying classNames to this component?

```js
class Menu extends React.Component {
  // ... data here
  render() {
    return (
      <nav className="menu">
        {this.props.menuItems.map((item) => (
          <div className="menu_buttonContainer">
            <button
              className={`menu_button ${
                this.state.active === item.id ? 'menu_Button-active' : ''
              }`}
            >
              {item.text}
            </button>
          </div>
        ))}
      </nav>
    );
  }
}
```

Not too bad. But what if we want to apply modifiers then it gets messy really quickly

```js
class Menu extends React.Component {
  // ... data here
  render() {
    return (
      <nav className="menu menu-right">
        {this.props.menuItems.map((item) => (
          <div className="menu_buttonContainer menu_buttonContainer-square">
            <button
              className={`menu_button ${
                this.state.active === item.id ? 'menu_button-active' : ''
              } ${this.state.animate === item.id ? 'menu_button-animate' : ''}`}
            >
              {item.text}
            </button>
          </div>
        ))}
      </nav>
    );
  }
}
```

Ughhh gross. You might try to clean it up with a helper function.

```js
class Menu extends React.Component {
  getButtonClass = (id) => {
    return `menu_button ${
      this.state.active === id ? 'menu_button-active' : ''
    } ${this.state.animate === id ? 'menu_button-animate' : ''}`;
  };
  // ... data here
  render() {
    return (
      <nav className="menu menu-right">
        {this.props.menuItems.map((item) => (
          <div className="menu_buttonContainer menu_buttonContainer-square">
            <button className={this.getButtonClass(item.id)}>
              {item.text}
            </button>
          </div>
        ))}
      </nav>
    );
  }
}
```

Well why not just use `<BEM>` ?

```js
class Menu extends React.Component {
  // same data from before

  state = {
    active: 1,
    animate: 1,
  };

  getModifiers = (id) => {
    return Object.entries(this.state)
      .filter((entry) => entry[1] === id)
      .map((entry) => entry[0]);
  };

  render() {
    return (
      <BEM block="menu">
        <nav modifier="right">
          {this.props.menuItems.map((item) => (
            <div element="buttonContainer" modifiers="square">
              <button element="button" modifiers={this.getModifiers(item.id)}>
                {item.text}
              </button>
            </div>
          ))}
        </nav>
      </BEM>
    );
  }
}
```

And that will produce the following html

```html
<nav modifier="right" class="menu">
    <div element="buttonContainer" modifiers="square" class="menu_buttonContainer menu_buttonContainer-square">
        <button element="button" modifiers="active,animate" class="menu_button menu_button-active menu_button-animate">Link1</button>
    </div>
    <div element="buttonContainer" modifiers="square" class="menu_buttonContainer menu_buttonContainer-square">
        <button element="button" modifiers="" class="menu_button">Link2</button>
    </div>
    <div element="buttonContainer" modifiers="square" class="menu_buttonContainer menu_buttonContainer-square">
        <button element="button" modifiers="" class="menu_button">Link3</button>
    </div>
</nav>
```

And now you can write your sass the same way

```scss
.menu
  &_buttonContainer
    &-square
  &_button
    &-animate
    &-active
```

Anyways you may think differently but this syntax makes alot of sense for me!

### Notes

BEM will only add classes to dom elements and not to Components. In the below example the p entities will receive classNames `"block_element"` and `"block_otherElement"` repectively. No classNames will be added to the `<SpecialComponent />`. This is a design feature not an error. We would assume that you will use another instance of BEM inside your special component as it is special enough to warrant it's own component! This frees you up for using the create rich environment of composing components together.

```js
render() {
  return (
    <BEM block="block">
      <p element="element">text</p>
      <p element="otherElement">text</p>
      <SpecialComponent />
    </BEM>
  )
}
```

All classNames are added to previously existing classNames. This means you can still use things like font awesome and existing classNames as BEM builds on top of already existing features. The i element below would have the className of `"fa block_element"`

```js
render() {
  return (
    <BEM block="block">
      <i className="fa" element="element">text</i>
    </BEM>
  )
}
```
