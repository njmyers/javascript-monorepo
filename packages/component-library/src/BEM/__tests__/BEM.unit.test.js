import React from 'react';
import BEM from '../BEM';

import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('the input works', () => {
  test('the input mounts without crashing', () => {
    const wrapper = shallow(<BEM />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('type check modifiers for string/array/undefined input', () => {
    const wrapper = shallow(<BEM />);
    const typeCheckModifiers = wrapper.instance().typeCheckModifiers;

    expect(typeof typeCheckModifiers).toBe('function');
    expect(typeCheckModifiers(undefined)).toMatchObject(['']);
    expect(typeCheckModifiers('')).toMatchObject(['']);
    expect(typeCheckModifiers('modifier')).toMatchObject(['', 'modifier']);
    expect(typeCheckModifiers(['modifier'])).toMatchObject(['', 'modifier']);
    expect(typeCheckModifiers(['modifier', 'active'])).toMatchObject([
      '',
      'modifier',
      'active',
    ]);
  });

  test('type check elements for string/undefined input', () => {
    const wrapper = shallow(<BEM />);
    const typeCheckElement = wrapper.instance().typeCheckElement;

    expect(typeof typeCheckElement).toBe('function');
    expect(typeCheckElement(undefined)).toBe('');
    expect(typeCheckElement('')).toBe('');
    expect(typeCheckElement('element')).toBe('element');
    expect(typeCheckElement('elementName')).toBe('elementName');
  });

  test('return correct BEM syntax -> getClassName method', () => {
    const wrapper = shallow(<BEM />);
    const getClassName = wrapper.instance().getClassName;

    expect(typeof getClassName).toBe('function');
    expect(getClassName('block')).toBe('block');
    expect(getClassName('block', 'element')).toBe('block_element');
    expect(getClassName('block', '', 'modifier')).toBe('block-modifier');
    expect(getClassName('block', 'element', 'modifier')).toBe(
      'block_element-modifier'
    );
    expect(getClassName('', 'element', 'modifier')).toBe('element-modifier');
  });

  test('getAllClasses method', () => {
    const wrapper = shallow(<BEM block="block" />);
    const getAllClasses = wrapper.instance().getAllClasses;

    expect(typeof getAllClasses).toBe('function');
    expect(getAllClasses('element', [''])).toBe('block_element');
    expect(getAllClasses('element', ['', 'modifier'])).toBe(
      'block_element block_element-modifier'
    );
    expect(getAllClasses('element', ['', 'modifier', 'red'])).toBe(
      'block_element block_element-modifier block_element-red'
    );

    expect(getAllClasses('', ['', 'modifier'])).toBe('block block-modifier');
    expect(getAllClasses('', ['', 'modifier', 'red'])).toBe(
      'block block-modifier block-red'
    );
  });

  test('mergeClassNames method with block, element and no modifier', () => {
    const wrapper = shallow(<BEM block="block" />);
    const mergeClassNames = wrapper.instance().mergeClassNames;

    const element = 'element';
    const modifiers = '';

    expect(typeof mergeClassNames).toBe('function');

    expect(mergeClassNames({ modifiers })).toMatchObject({
      className: 'block',
    });

    expect(mergeClassNames({ element, modifiers })).toMatchObject({
      className: 'block_element',
    });
  });

  test('mergeClassNames method with block and single modifier', () => {
    const wrapper = shallow(<BEM block="block" />);
    const mergeClassNames = wrapper.instance().mergeClassNames;

    const element = 'element';
    const modifiers = 'modifier';

    expect(typeof mergeClassNames).toBe('function');

    expect(mergeClassNames({ modifiers })).toMatchObject({
      className: 'block block-modifier',
    });

    expect(mergeClassNames({ element, modifiers })).toMatchObject({
      className: 'block_element block_element-modifier',
    });
  });

  // test('it renders a child element', () => {
  //   const wrapper = shallow(
  //     <BEM>
  //       <div>Some Content</div>
  //     </BEM>
  //   );

  //   console.log(wrapper.find('div').props());

  //   expect(wrapper.length).toBe(1);
  //   expect(wrapper.find('div').exists()).toBe(true);
  // });

  // test('it applies a block className to a child element', () => {
  //   const wrapper = shallow(
  //     <BEM block="block">
  //       <div>Some Content</div>
  //     </BEM>
  //   );

  //   expect(wrapper.find('div').props().className).toBe('block');
  // });

  // test('the input renders default classNames', () => {
  //   const wrapper = shallow(<Input />);
  //   expect(wrapper.find('div').props().className).toBe('formInput');
  //   expect(wrapper.find('label').props().className).toBe('formInput_label');
  //   expect(wrapper.find('input').props().className).toBe('formInput_input');
  // });

  // test('the input renders input classNames', () => {
  //   const wrapper = shallow(<Input className="addressInput" />);
  //   expect(wrapper.find('div').props().className).toBe('addressInput');
  //   expect(wrapper.find('label').props().className).toBe('addressInput_label');
  //   expect(wrapper.find('input').props().className).toBe('addressInput_input');
  // });

  // test('the input renders adds modifier classNames', () => {
  //   const wrapper = shallow(
  //     <Input className="addressInput" classModifier="active" />
  //   );
  //   expect(wrapper.find('div').props().className).toBe(
  //     'addressInput addressInput-active'
  //   );
  //   expect(wrapper.find('label').props().className).toBe(
  //     'addressInput_label addressInput_label-active'
  //   );
  //   expect(wrapper.find('input').props().className).toBe(
  //     'addressInput_input addressInput_input-active'
  //   );
  // });

  // test('the input renders adds multiple modifier classNames', () => {
  //   const wrapper = shallow(
  //     <Input className="addressInput" classModifier={['active', 'red']} />
  //   );
  //   expect(wrapper.find('div').props().className).toBe(
  //     'addressInput addressInput-active addressInput-red'
  //   );
  // });
});
