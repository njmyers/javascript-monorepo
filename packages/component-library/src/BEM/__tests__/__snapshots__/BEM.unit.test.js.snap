// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the input works the input mounts without crashing 1`] = `null`;
